src/server/dflycmd.cc

```diff
@@ -164,39 +203,87 @@ void DflyCmd::Thread(CmdArgList args, ConnectionContext* cntx) {
   RedisReplyBuilder* rb = static_cast<RedisReplyBuilder*>(cntx->reply_builder());
-  string_view masterid = ArgS(args, 2);
+  string_view master_id = ArgS(args, 2);
   string_view sync_id_str = ArgS(args, 3);
   string_view flow_id_str = ArgS(args, 4);
 
-  unsigned flow_id;
-  VLOG(1) << "Got DFLY SYNC " << masterid << " " << sync_id_str << " " << flow_id_str;
+  VLOG(1) << "Got DFLY FLOW " << master_id << " " << sync_id_str << " " << flow_id_str;
 
-  if (masterid != sf_->master_id()) {
-    return rb->SendError("Bad master id");
+  if (master_id != sf_->master_id()) {
+    return rb->SendError(kBadMasterId);
   }
 
-  if (!absl::SimpleAtoi(flow_id_str, &flow_id) || !absl::StartsWith(sync_id_str, "SYNC")) {
-    return rb->SendError(kSyntaxErr);
+  unsigned flow_id;
+  if (!absl::SimpleAtoi(flow_id_str, &flow_id) || flow_id >= shard_set->pool()->size()) {
+    return rb->SendError(facade::kInvalidIntErr);
   }
 
   auto [sync_id, sync_info] = GetSyncInfoOrReply(sync_id_str, rb);
   if (!sync_id)
     return;
 
-  // assuming here that shard id and thread id is the same thing.
-  if (int(flow_id) != ProactorBase::GetIndex()) {
-    listener_->Migrate(cntx->owner(), shard_set->pool()->at(flow_id));
+  unique_lock lk(sync_info->mu);
+  if (sync_info->state != SyncState::PREPARATION)
+    return rb->SendError(kInvalidState);
+
+  // Set meta info on connection.
+  cntx->owner()->SetName(absl::StrCat("repl_flow_", sync_id));
+  cntx->conn_state.repl_session_id = sync_id;
+  cntx->conn_state.repl_flow_id = flow_id;
+
+  absl::InsecureBitGen gen;
+  string eof_token = GetRandomHex(gen, 40);
+
+  sync_info->flows[flow_id] = FlowInfo{cntx->owner(), eof_token};
+  listener_->Migrate(cntx->owner(), shard_set->pool()->at(flow_id));
+
+  rb->StartArray(2);
+  rb->SendSimpleString("FULL");
+  rb->SendSimpleString(eof_token);
+}
+
+void DflyCmd::Sync(CmdArgList args, ConnectionContext* cntx) {
+  RedisReplyBuilder* rb = static_cast<RedisReplyBuilder*>(cntx->reply_builder());
+  string_view sync_id_str = ArgS(args, 2);
+
+  VLOG(1) << "Got DFLY SYNC " << sync_id_str;
+
+  auto [sync_id, sync_info] = GetSyncInfoOrReply(sync_id_str, rb);
+  if (!sync_id)
+    return;
+
+  unique_lock lk(sync_info->mu);
+  if (sync_info->state != SyncState::PREPARATION)
+    return rb->SendError(kInvalidState);
+
+  // Check all flows are connected.
+  for (const FlowInfo& flow : sync_info->flows) {
+    if (!flow.conn) {
+      return rb->SendError(kInvalidState);
+    }
   }
 
-  (void)sync_id;
-  (void)sync_info;
+  // Start full sync.
+  {
+    TransactionGuard tg{cntx->transaction};
+    AggregateStatus status;
+
+    auto cb = [this, &status, sync_info = sync_info](unsigned index, auto*) {
+      status = StartFullSyncInThread(&sync_info->flows[index], EngineShard::tlocal());
+    };
+    shard_set->pool()->AwaitFiberOnAll(std::move(cb));
 
+    // TODO: Send better error
+    if (*status != OpStatus::OK)
+      return rb->SendError(kInvalidState);
+  }
```

Comment:
We have the transaction guard + shard_set dispatch instead of a regular transaction because we need to run on the non-shard thread as well.

Any better ideas?